### ISSUE TITLE
Assessing error of convnet method

### ISSUE THREAD

**Issue by DABH** [fluidnetsc22/fluidnet_sc22#2]
Hi @fluidnetsc22 ,

Thanks so much for publishing this repo and the excellent work your team has done for this project.  I'm sure it will be an interesting paper :)

When using the convnet method (i.e., FluidNet) for performing the simulation, is there an easy way with your code to see what the error is, i.e. `Ax-b` or `||Ax-b||` for the pressure Poisson system that's being solved?  I read through the code, e.g. `lib/simulate.py` around line 296, and I see that we get the `p` (`x`) and have the `div` (`b`), but the Poisson matrix is never explicitly formed, and it's not 100% clear what the boundary conditions are and how the matrix is ordered and scaled.  So, if there isn't some functionality for computing the residual vector/norm, could you provide some code for explicitly forming `A`?  Many thanks in advance for any help or code you can provide!

**Comment by fluidnetsc22** [comment]
Hi @DABH,

First, thank you very much for taking the time for looking around the code, I really appreciate it as all the possible comments and remarks help us improve the quality and clearness of our work.

Concerning the error evaluation, there are two answers that come up to my mind:

1. To evaluate the residual Ax-b one can look at the corrected velocity divergence, as it corresponds to a direct measure of the residual. The Poisson equation solved by the networks is `\alpha \nabla^2 p = \nabla \cdot u^*` (which is basically the AX=b),  where the residual (AX-b) can be written as  \alpha \nabla \cdot u^* - \nabla^2 p.  However, it should be noted that this equation is computed from taking the divergence of both sides of the pressure update step (u = u^* - \alpha \nabla^2 p), and its complete form is :  \nabla \cdot u =   \nabla \cdot u^* - \alpha \nabla^2 p, where \nabla \cdot u should be equal to 0 if the equation is well solved. Thus, the residual (AX-b) is just equal to the divergence of the corrected velocity field \nabla \cdot u . Evaluating this divergence should be analogous to evaluation of the residual AX-b.

2. Concerning the A matrix, it is true that when solving with the neural network, it is not explicitly formed. However, the matrix is indeed created when using the Conjugate Gradient method (for example, in the VK test case). Matrix A is computed from the occupancy matrix (or flags matrix), for example, in line 276 of the [VK script](https://github.com/fluidnetsc22/fluidnet_sc22/blob/main/pytorch/VK_3D.py). The A matrix is directly stored in CSR format, using the script found [here](https://github.com/fluidnetsc22/fluidnet_sc22/blob/main/pytorch/lib/fluid/MatrixA_3D.py).

I hope that these answers help clarify your concerns, but do not hesitate to ask any further questions or clarifications on the matter.

Thanks in advance,

fluidnetsc22

**Comment by DABH** [comment]
Thank you so much for your very fast and helpful reply!  Your explanations are very clear, and this helped me get everything working (I just used div U as the residual, and was able to copy the code from the VK script to get the matrix in the smoke plume example).  Many thanks again for the wonderful support! :)
